src/libs/libpng/lv_libpng.c

```diff
@@ -0,0 +1,302 @@
+        /*Save the data in the header*/
+        header->always_zero = 0;
+        header->cf = LV_COLOR_FORMAT_ARGB8888;
+        /*The width and height are stored in Big endian format so convert them to little endian*/
+        header->w = (lv_coord_t)((size[0] & 0xff000000) >> 24) + ((size[0] & 0x00ff0000) >> 8);
+        header->h = (lv_coord_t)((size[1] & 0xff000000) >> 24) + ((size[1] & 0x00ff0000) >> 8);
+
+        return LV_RES_OK;
+    }
+
+    return LV_RES_INV;         /*If didn't succeeded earlier then it's an error*/
+}
+
+
+/**
+ * Open a PNG image and return the decided image
+ * @param src can be file name or pointer to a C array
+ * @param style style of the image object (unused now but certain formats might use it)
+ * @return pointer to the decoded image or  `LV_IMAGE_DECODER_OPEN_FAIL` if failed
+ */
+static lv_res_t decoder_open(lv_image_decoder_t * decoder, lv_image_decoder_dsc_t * dsc)
+{
+    LV_UNUSED(decoder); /*Unused*/
+
+    /*If it's a PNG file...*/
+    if(dsc->src_type == LV_IMAGE_SRC_FILE) {
+        const char * fn = dsc->src;
+        lv_cache_lock();
+        lv_cache_entry_t * cache = lv_cache_add(dsc->header.w * dsc->header.h * sizeof(uint32_t));
+        if(cache == NULL) {
+            lv_cache_unlock();
+            return LV_RES_INV;
+        }
+
+        uint32_t t = lv_tick_get();
+        const void * decoded_img = decode_png_file(fn);
+        t = lv_tick_elaps(t);
+        cache->weight = t;
+        cache->data = decoded_img;
+        cache->free_data = 1;
+        if(dsc->src_type == LV_IMAGE_SRC_FILE) {
+            cache->src = lv_strdup(dsc->src);
+            cache->src_type = LV_CACHE_SRC_TYPE_STR;
+            cache->free_src = 1;
+        }
+        else {
+            cache->src_type = LV_CACHE_SRC_TYPE_PTR;
+            cache->src = dsc->src;
+        }
+
+        dsc->img_data = lv_cache_get_data(cache);
+        dsc->user_data = cache;
+
+        lv_cache_unlock();
+        return LV_RES_OK;     /*The image is fully decoded. Return with its pointer*/
+    }
+
+    return LV_RES_INV;    /*If not returned earlier then it failed*/
+}
+
+/**
+ * Free the allocated resources
+ */
+static void decoder_close(lv_image_decoder_t * decoder, lv_image_decoder_dsc_t * dsc)
+{
+    LV_UNUSED(decoder); /*Unused*/
+
+    lv_cache_lock();
+    lv_cache_release(dsc->user_data);
+    lv_cache_unlock();
+}
+
+static lv_result_t try_cache(lv_image_decoder_dsc_t * dsc)
+{
+    lv_cache_lock();
+    if(dsc->src_type == LV_IMAGE_SRC_FILE) {
+        const char * fn = dsc->src;
+
+        lv_cache_entry_t * cache = lv_cache_find(fn, LV_CACHE_SRC_TYPE_STR, 0, 0);
+        if(cache) {
+            dsc->img_data = lv_cache_get_data(cache);
+            dsc->user_data = cache;     /*Save the cache to release it in decoder_close*/
+            lv_cache_unlock();
+            return LV_RESULT_OK;
+        }
+    }
+
+    else if(dsc->src_type == LV_IMAGE_SRC_VARIABLE) {
```

Comment:
It seems in `decoder_open` only files are handled, so the `if(dsc->src_type == LV_IMAGE_SRC_VARIABLE)` shouldn't be required.